examples_tests/15.LayeredCubeMapShadows/deriv_map_gen.comp

```diff
@@ -1,97 +1,46 @@
-		return subgroupClusteredAdd(intBitsToFloat(fbits ^ xormask), 2);
-	}
-	#endif
-#endif
-#endif //#if 0 
-//possible use of extensions is turned off for now because untested due to:
-//1) my GTX660 doesn't support GL_NV_compute_shader_derivatives (actually seems like only high-end RTXs support this)
-//2) weird error "unknown builtin varying" concerning gl_SubgroupInvocationID even though GL_KHR_shader_subgroup_clustered is supported (driver bug?)
-
-// also distribution of work-groups on dimensions should depend on available GLSL extensions
-#if defined(IRR_DDX_IS_DEFINED) && defined(IRR_DDY_IS_DEFINED)
-layout (local_size_x = 16, local_size_y = 16) in;
-#else
-	#if XPASS
-	layout (local_size_x = 256) in;
-	#else
-	layout (local_size_y = 256) in;
-	#endif
-#endif
-
-#if !defined(IRR_DDX_IS_DEFINED) && !defined(IRR_DDY_IS_DEFINED)
-shared float smem[256]; //1KB
-#endif
-
-layout (binding = 7) uniform sampler2D bumpMapSampler;
-#if !defined(IRR_DDY_IS_DEFINED)
-layout (binding = 0, r16_snorm) uniform image2D derivativeMapImage;
-#else
-layout (binding = 0, rg16_snorm) uniform image2D derivativeMapImage;
-#endif
-
-layout (location = 0) uniform float uHeightScaleFactor = 0.1;
 
 void main()
 {
 	const ivec2 G_IDX = ivec2(gl_GlobalInvocationID.xy);
 	const ivec2 RESOLUTION = ivec2(gl_NumWorkGroups.xy * gl_WorkGroupSize.xy);
 	
-#if defined(IRR_DDX_IS_DEFINED) && defined(IRR_DDY_IS_DEFINED)
-	float smpl = textureLod(bumpMapSampler, vec2(1.0, 1.0)/vec2(RESOLUTION) * vec2(G_IDX), 0.0).r * uHeightScaleFactor;
-	float dx = DDX(smpl);
-	float dy = DDY(smpl);
-	
-	if ((G_IDX.x & 1) == 0 && (G_IDX.y & 1) == 0)
-		imageStore(derivativeMapImage, G_IDX/2, vec4(dx, dy, 0.0, 0.0));
-#elif defined(IRR_DDX_IS_DEFINED)
-	float d = DDX(textureLod(bumpMapSampler, vec2(1.0, 1.0)/vec2(RESOLUTION) * vec2(G_IDX), 0.0).r * uHeightScaleFactor);
-	if ((G_IDX.x & 1) == 0 && (G_IDX.y & 1) == 0)
-		imageStore(derivativeMapImage, G_IDX/2, vec4(d, 0.0, 0.0, 0.0));
-#else
-	const uint LC_IDX = gl_LocalInvocationIndex;
-	smem[LC_IDX] = textureLod(bumpMapSampler, vec2(1.0, 1.0)/vec2(RESOLUTION) * vec2(G_IDX), 0.0).r * uHeightScaleFactor;
+	vec2 bumpMapSz = vec2(textureSize(bumpMapSampler, 0));
+
+	const ivec2 LC_IDX = ivec2(gl_LocalInvocationID.xy);
+	smem[getAddr(LC_IDX)] = textureLod(bumpMapSampler, vec2(1.0, 1.0)/vec2(RESOLUTION) * vec2(G_IDX), 0.0).r * uHeightScaleFactor;
+	if (LC_IDX.x == 0 || LC_IDX.x == 15)
+	{
+		ivec2 offset = ivec2(mix(ivec2(1, 0), ivec2(-1, 0), LC_IDX.x == 0));
+		smem[getAddr(LC_IDX+offset)] = textureLod(bumpMapSampler, vec2(1.0, 1.0)/vec2(RESOLUTION) * vec2(G_IDX+offset), 0.0).r * uHeightScaleFactor;
+	}
+	if (LC_IDX.y == 0 || LC_IDX.y == 15)
+	{
+		ivec2 offset = ivec2(mix(ivec2(0, 1), ivec2(0, -1), LC_IDX.y == 0));
+		smem[getAddr(LC_IDX+offset)] = textureLod(bumpMapSampler, vec2(1.0, 1.0)/vec2(RESOLUTION) * vec2(G_IDX+offset), 0.0).r * uHeightScaleFactor;
+	}
```

Comment:
also why not apply `uHeightScaleFactor` at the end?

Reply:
because now it works on smaller numbers = possibly better precision. Generally i think that way, but here no point

Reply:
float precision is only an issue if your operands to ADD or SUB differ vastly in magnitude